src/com/walmartlabs/lacinia/pedestal/subscriptions.clj

```diff
@@ -102,77 +85,204 @@
+                   "connection_init"
+                   (when (>! response-data-ch {:type :connection_ack})
+                     (recur (assoc connection-state :connection-params payload)))
+
+                   ;; TODO: Track state, don't allow start, etc. until after connection_init
+
+                   "start"
+                   (if (contains? (:subs connection-state) id)
+                     (do
+                       (log/debug :event ::ignoring-duplicate :id id)
+                       (recur connection-state))
+                     (do
+                       (log/debug :event ::start :id id)
+                       (let [merged-context (assoc context :connection-params (:connection-params connection-state))
+                             sub-shutdown-ch (execute-query-interceptors id payload response-data-ch cleanup-ch merged-context)]
+                         (recur (assoc-in connection-state [:subs id] sub-shutdown-ch)))))
+
+                   "stop"
+                   (do
+                     (log/debug :event ::stop :id id)
+                     (when-some [sub-shutdown-ch (get-in connection-state [:subs id])]
+                       (close! sub-shutdown-ch))
+                     (recur connection-state))
+
+                   "connection_terminate"
+                   (do
+                     (log/debug :event ::terminate)
+                     (run! close! (-> connection-state :subs vals))
+                     ;; This shuts down the connection entirely.
+                     (close! response-data-ch))
+
+                   ;; Not recognized!
+                   (let [response (cond-> {:type :error
+                                           :payload {:message "Unrecognized message type."
+                                                     :type type}}
+                                          id (assoc :id id))]
+                     (log/debug :event ::unknown-type :type type)
+                     (>! response-data-ch response)
+                     (recur connection-state))))))))))
+    (handle-parse-error [_ _ t response-data-ch]
+      (put! response-data-ch {:type :connection_error
+                              :payload (util/as-error-map t)}))
+    (->data-message [_ id response] {:type :data
+                                     :id id
+                                     :payload response})
+    (->complete-message [_ id] {:type :complete
+                                :id id})
+    (->error-message [_ id payload] {:type :error
+                                     :id id
+                                     :payload payload})))
+
+(def ^:private graphql-transport-ws
+  "Server side implementation for `graphql-transport-ws` sub-protocol.
+  specified by [graphql-ws](https://github.com/enisdenjo/graphql-ws/blob/v5.7.0/PROTOCOL.md)"
+  (reify Ws-Sub-Protocol
+    (sub-protocol-name [_] "graphql-transport-ws")
+    (connection-loop [self session _ ws-data-ch response-data-ch context]
```

Comment:
Really not using self here.  Also, from bitter experience, I tend to have methods in reified objects delegate to a function; otherwise you hit some oddities about debugging and reloading.